### ISSUE TITLE
Dangerous memory leak on xbmc.executeJSONRPC affects Kodi 18 and 19

### ISSUE THREAD

**Issue by CastagnaIT** [xbmc/xbmc#19332]
### Describe the bug
Here is a clear and concise description of what the problem is:

Some users had report to me a strange behaviour of AddonSignals module related to some Kodi crashes
what happens?

the heart of AddonSignals operation is to use `xbmc.executeJSONRPC` as way to exchange data, the problem is that
when small data are sent many times, no deterioration is noticed to the RAM,
when sending large amounts of data, the device/computer starts to suffer, little by little the RAM will collapse,
somewhere the data is not released.

Simply trying to execute many times `xbmc.executeJSONRPC` with `JSONRPC.NotifyAll` assigning some MB of text data,
it will eat up all the RAM on the device until Kodi crashes.

Sidenote: i do not know if happen with all JSONRPC commands or only with `JSONRPC.NotifyAll` that have the data parameter.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Use xbmc.executeJSONRPC without memory leak

## Actual Behavior

An image of mem leak, almost 10GB or RAM allocated and never released:
![immagine](https://user-images.githubusercontent.com/3257156/110002518-c3eff400-7d15-11eb-9034-7e37819b0934.png)
(50 AddonSignals make_calls, where 10MB text trasmitted + 10MB text in the response)

### To Reproduce

1. Install the test addon: 
[plugin.video.test.zip](https://github.com/xbmc/xbmc/files/6085544/plugin.video.test.zip)
2. Open Task Manager select Kodi
2. Run the addon
3. See that Kodi process have eated ~3GB of RAM

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/lequvipopi.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10
 - Kodi version: Kodi 19 (happen also on Kodi 18)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by enen92** [comment]
Hey @CastagnaIT,
Thanks for yet exceptional bug report (and support addon). Can you check if the following patch fixes the issue?

```
diff --git a/xbmc/interfaces/python/XBPython.cpp b/xbmc/interfaces/python/XBPython.cpp
index a02b7d4415..fca86a4eaf 100644
--- a/xbmc/interfaces/python/XBPython.cpp
+++ b/xbmc/interfaces/python/XBPython.cpp
@@ -419,6 +419,7 @@ void XBPython::OnNotification(const std::string& sender,
     if (CHECK_FOR_ENTRY(m_vecMonitorCallbackList, it))
       it->OnNotification(sender, method, data);
   }
+  tmp.clear();
 }

 void XBPython::Uninitialize()
```

I checked the mem usage keeps constant but didn't exactly test python to see if we get any regression there. I doubt but could use some testing :)

**Comment by CastagnaIT** [comment]
hi @enen92 thanks, provide a test addon speeds things up
i have tried build your changes
and also tried a second build by adding also `lock.Leave();` but same thing, the results are the same:

by using test add-on,
before (no patch) ~4GB allocated (not released), after the patch ~2GB still allocated (not released),
about 50% of improvement but somewhere continues there is still unreleased memory

**Comment by CastagnaIT** [comment]
i have see that if i add `jsonData.clear();` after `if CJSONVariantWriter::Write` execution:
https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/python/XBPython.cpp#L110-L118

and plus adding your patch to each methods with `LOCK_AND_COPY` seem to FULL RESOLVE the problem
but i do not know if add `jsonData.clear();` in that place is a right thing to do please check

**Comment by CastagnaIT** [comment]
I correct myself something else not works well
with my last changes with `jsonData.clear();`
there is a strange behaviour on this memory release

you can check by apply my changes and modify the test add-on by changing the:
`while i<50:` to `while i<200:`

you will see that memory is allocated but not immediately released
and causes a visible delay in closing the GUI progress bar window

however the problem is around here

**Comment by enen92** [comment]
Yeah, my patch doesn't do anything, it was a brainfart from my side :) Obviously the `tmp` vector is deleted whenever the `OnNotification` method goes out of scope.

I took a further look into this yesterday night but was already too late to answer. After playing a bit with this I don't think we have any memory leak. The problem happens because, as all of this is asynchronous, the rate at which you add new elements to the queue (https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/AnnouncementManager.cpp#L124-L144) is much greater than the rate at which you are able to take the events out of the queue and send the notifications.
A new thread is spanned each time you add a new notification via json but the thread which actually sends the events is running and waiting for the event to be set (https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/AnnouncementManager.cpp#L313-L335). So, in between actually sending the events, the producer can lock the queue and add new ones.

My only idea is to make sure we empty the queue any time we process the messages. This actually helps lowering the memory usage and will probably work for you:

```cpp
void CAnnouncementManager::Process()
{
  SetPriority(GetMinPriority());

  while (!m_bStop)
  {
    CSingleLock lock (m_queueCritSection);
    if (!m_announcementQueue.empty())
    {
     while (!m_announcementQueue.empty())
     {
      auto announcement = m_announcementQueue.front();
      m_announcementQueue.pop_front();
      {
        CSingleExit ex(m_queueCritSection);
        DoAnnounce(announcement.flag, announcement.sender.c_str(), announcement.message.c_str(), announcement.item, announcement.data);
      }
     }
    }
    else
    {
      CSingleExit ex(m_queueCritSection);
      m_queueEvent.Wait();
    }
  }
``` 

Maybe @fuzzard or @phunkyfish may have better ideas or a known pattern to solve this

**Comment by CastagnaIT** [comment]
i have investigated by removing and skipping many parts of Kodi code, and i think i found the cause
or at least i found the zone of the problematic code

i have try done a fix, open my commit:
https://github.com/CastagnaIT/xbmc/commit/7ae641077d9c5e906eb563026f8b7abc550ee643

and look the changes on file:
- xbmc/interfaces/legacy/Monitor.h

the object `CallbackFunction` created previously on the fly, seem that is not released from memory
with my commit changes, the results are:
~200 mb RAM occupied and full released at the end of the operations instead of ~3GB of RAM

but my change is not perfect because i noticed some crashes for memory access,
that i have fixed in some way on the file:
- xbmc/interfaces/python/XBPython.cpp

need to look better

**Comment by CastagnaIT** [comment]
I have reworked my memory leak fix,
because the hole is right where i suspected
and this time the fix works without crashes

as mentioned initially i have encoutered the problem in Monitor.h where a new object is instantiated and passed as a pointer to `invokeCallback`, after this, we have seen that the memory allocated will never be released
this because i think the `delete` is never called, but also if you call delete, will cause crashes, maybe because there is various memory references?

however this cause a **big memory leak in whole kodi core** because `invokeCallback` method is called in many parts of Kodi, not only from `xbmc.executeJSONRPC` which makes use of `onNotification` only.

This is my changes to prove how I fixed the memory leak:
https://github.com/CastagnaIT/xbmc/commit/4fb044c60bf212f6704ec0e07958243f9175f170
(i have applied it only in specific for `OnNotification` method)

in my commit i have done a custom implementation of `invokeCallback` that now make use of `std::shared_ptr`
in the same way i have readapted `CallbackHandler.cpp` for the use of `std::shared_ptr`
not completely because I did not understand how to create a sharedptr from an interface (see RetardedAsyncCallbackHandler)
in any case this is for show the things

These are the the results after running 3 times the test addon:
- Memory occopied max ~700mb
- Memory fully released at the end of add-on operations (Kodi standard on windows64 about 100mb same on fresh kodi startup)

this proves that Kodi currently has a serious problem of memory leak that continues to be ignored
my example of generating large volumes of MB is just the best way to see the problem

this means that any device where Kodi is running
before or after will cause a system crash due to memory occupied and never released over the time

**Comment by CastagnaIT** [comment]
@repojohnray i see that you are more or less expert on investigate for memory leaks
if you have time can you try investigate to understand what happens here?

**Comment by basilgello** [comment]
@CastagnaIT we need to instrument Kodi with Valgrind to find the clues. Just note building Kodi with Valgrind needs to have tests disabled if you dont want wasting hours. Let me find my notes enabling Valgrind in Kodi.

**Comment by repojohnray** [comment]
@CastagnaIT, this could be related to m_types (std::map) which prevents the std::shared_ptr object (JSONSchemaTypeDefinitionPtr) to be freed. Once JSONRPC::CJSONRPC::Cleanup() (NetworkServices.cpp) is called a lot of these objects are removed from memory.
